services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py

```diff
@@ -1,68 +1,203 @@
+import asyncio
 import json
 import logging
 import re
-from datetime import datetime
+from datetime import datetime, timedelta, timezone
 
 from fastapi import FastAPI
+from models_library.generated_models.docker_rest_api import Availability, Node, Task
 from pydantic import parse_obj_as
 from types_aiobotocore_ec2.literals import InstanceTypeType
 
 from ._meta import VERSION
 from .core.errors import Ec2InstanceNotFoundError
 from .core.settings import ApplicationSettings
+from .models import Resources
 from .modules.docker import get_docker_client
-from .modules.ec2 import get_ec2_client
+from .modules.ec2 import EC2InstanceData, get_ec2_client
 from .utils import ec2, rabbitmq, utils_docker
 
 logger = logging.getLogger(__name__)
 
 _EC2_INTERNAL_DNS_RE: re.Pattern = re.compile(r"^(?P<ip>ip-[0-9-]+).+$")
 
 
-async def check_dynamic_resources(app: FastAPI) -> None:
-    """Check that there are no pending tasks requiring additional resources in the cluster (docker swarm)
-    If there are such tasks, this method will allocate new machines in AWS to cope with
-    the additional load.
-    """
+async def _mark_empty_active_nodes_to_drain(
+    app: FastAPI,
+    monitored_nodes: list[Node],
+) -> None:
     app_settings: ApplicationSettings = app.state.settings
     assert app_settings.AUTOSCALING_NODES_MONITORING  # nosec
-
-    # 1. get monitored nodes information and resources
     docker_client = get_docker_client(app)
-    monitored_nodes = await utils_docker.get_monitored_nodes(
-        docker_client,
-        node_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_NODE_LABELS,
+    active_empty_nodes = [
+        node
+        for node in monitored_nodes
+        if (
+            await utils_docker.compute_node_used_resources(
+                docker_client,
+                node,
+                service_labels=app_settings.AUTOSCALING_NODES_MONITORING.NODES_MONITORING_SERVICE_LABELS,
+            )
+            == Resources.create_as_empty()
+        )
+        and (node.Spec is not None)
+        and (node.Spec.Availability == Availability.active)
```

Comment:
so this is an interesting thought. What might happen that the docker swarm reports a node as down while it is still running?
1. networking problem within AWS after a successful initial connection
2. the machine crashed/hanged/non responsive
3. something else?

At the moment, the code will just remove any down node from the docker swarm nodes list, to prevent it from growing. it will not terminate the instance (with the current code that is). Now this is maybe a bad idea, I could also wait some time before removing it (such as checking the last updated field on the node, and maybe let an hour go through or similar).

Doing such a thing could maybe solve 1., 2. if someone manages to restart the broken machine within the timeout period. But as of now, that is beyond the scope of that service I think. we can brainstorm what might come as additional features.

Reply:
created a new issue out of this https://github.com/ITISFoundation/osparc-simcore/issues/3684